hooks/useModel.ts

```diff
@@ -1,67 +1,194 @@
+     * If no names are given then all model values are reset.
+     */
+    $reset(...names: Array<keyof T>): void
+}
 
-/**
- * Model object structure.
- */
-export type Model<T extends object, K extends keyof T = keyof T> = {
-    [key in K]: {
-        value: any
-        onChange: (input: ModelInputValue) => void
-    }
-} & {$data: T; $set: ReactStateSetter<Partial<T> | T>}
-/**
- * Valid model inputs values.
- */
-export type ModelInputValue = ParseableInput | ModelPrimitive
-type ModelPrimitive = string | boolean | number | object
-
-/**
- * Creates an input model.
- *
- * @param init The initial model object data.
- * @param middleware A function that is called right before updating the model.
- * This allows for final custom manipulations on the value.
- */
 function useModel<T extends object>(
-    init: T,
-    middleware?: (val: ModelPrimitive, name: string) => ModelPrimitive | void
+    initialElements: () => Partial<
+        Dictionary<T, string | boolean | number | null | ModelElement<T>>
+    >
 ): Model<T> {
-    const [data, setModel] = useState(init)
+    const [state, dispatch] = useReducer(modelHookReducer, {}, initialize)
 
-    function change(name: string, input: ModelInputValue) {
-        let value = input
-        if (
-            isType<ParseableInput>(input, v => v.target || isObject(v, Element))
-        )
-            value = cast(parseInputValue(input))
+    function initialize() {
+        const elements = initialElements()
 
-        value = call(middleware, value, name) || value
+        const utils: ModelUtils = {}
+        const model = {
+            $change(values: ModelValueList<any>) {
+                dispatch(modelChangeAction(values))
+            },
+            $data() {
+                const values: any = {}
+                for (const name in utils) {
+                    values[name] = this[name].value
+                }
+                return values
+            },
+            $errors() {
+                const data = this.$data()
+                const errors: any = {}
+                for (const name in utils) {
+                    const entry = this[name]
+                    const err = call(utils[name].validate, entry.value, data)
+                    if (err) {
+                        errors[name] = err
+                    }
+                }
+                return errors
+            },
+            $firstError() {
+                return Object.values(this.$errors())[0]
+            },
+            $reset(...names: string[]) {
+                const nameList = len(names) ? names : Object.keys(elements)
+                const values: any = {}
 
-        setModel({
-            ...data,
-            [name]: value,
-        })
-    }
+                for (const name of nameList) {
+                    // @ts-ignore Element implicitly has an 'any' type... (7017)
+                    const current = elements[name]
+                    values[name] = isObject<ModelElement>(current)
```

Comment:
Should NOT assign undefined values and instead assign an empty string to undefined ones. __Consider doing this for all undefined values in the reducer.__
```suggestion
                    value = isObject(...)
                    values[name] = value === undefined ? "" : value
```